### PR TITLE
cvv.py: Support multi-release JAR files

### PR DESCRIPTION
**Pull request by the-horo**
Closes: https://bugs.gentoo.org/900767

### CODE REVIEW

**Comment by the-horo** [comment]
@vaukai

File: src/javatoolkit/cvv.py [hunk 1]
```diff
@@ -19,10 +19,13 @@ def __init__(self, target):
         self.bad = []
         self.skipped = []
 
-    def add(self, version, jar, file):
-        if file == "module-info.class" and self.target < 9:
+    def add(self, version, jar, file, target_version: T.Optional[int] = None):
+        if target_version is None:
+            target_version = self.target
+
+        if file == "module-info.class" and target_version < 9:
```
Comment (vaukai) [inline src/javatoolkit/cvv.py]: multi-release JAR files need not necessarily have `module-info`.

**Comment by vaukai** [comment]
> @vaukai

@the-horo 

1. Sorry, but I am not someone to review the coding.  Is there a PR or a branch in `::gentoo` which I could use for testing?

**Comment by the-horo** [comment]
> > @vaukai
> 
> @the-horo
> 
>     1. Sorry, but I am not someone to review the coding.  Is there a PR or a branch in `::gentoo` which I could use for testing?

No, you would need a `javatoolkit-9999.ebuild` and point it to my fork or a local copy that contains my branch: https://github.com/the-horo/javatoolkit/tree/multi-release-jars. Here's the ebuild if it helps you:
```ebuild
# Copyright 1999-2025 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

EAPI=8

PYTHON_COMPAT=( python3_{10..12} )
PYTHON_REQ_USE="xml(+)"
DISTUTILS_USE_PEP517=flit

inherit distutils-r1 prefix

DESCRIPTION="Collection of Gentoo-specific tools for Java"
HOMEPAGE="https://wiki.gentoo.org/wiki/Project:Java"
if [[ ${PV} == *9999* ]]; then
	EGIT_REPO_URI="https://anongit.gentoo.org/git/proj/javatoolkit.git"
	inherit git-r3
else
	SRC_URI="https://gitweb.gentoo.org/proj/${PN}.git/snapshot/${P}.tar.bz2"
	KEYWORDS="~amd64 ~arm64 ~ppc64 ~sparc ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x64-solaris"
fi

LICENSE="GPL-2"
SLOT="0"

python_prepare_all() {
	hprefixify src/javatoolkit/scripts/findclass.py
	distutils-r1_python_prepare_all
}

python_install() {
	distutils-r1_python_install

	# The java eclasses expect the scripts to be in a special location
	python_scriptinto /usr/libexec/${PN}
	python_doexe "${D}$(python_get_scriptdir)"/*
}
```

File: src/javatoolkit/cvv.py (hunk 1 above)
Comment (the-horo) [reply src/javatoolkit/cvv.py]: I'm not sure how this is relevant. If some version subdirectory in a JAR file doesn't contain a module-info the rest of the files would still be checked normally, i.e. that their bytecode encoded version is less or equal to the version of the subdirectory.

There is a bug however, the `file == "module-info.class"` checks is technically wrong as the filename in a multi-release jar would be something like `META-INF/versions/9/module-info.class` though practically the rest of the condition (`target_version < 9`) is false because `do_jar` skips versions less than 9.

**Comment by vaukai** [comment]
> > > @vaukai
> > 
> > 
> > @the-horo
> > ```
> > 1. Sorry, but I am not someone to review the coding.  Is there a PR or a branch in `::gentoo` which I could use for testing?
> > ```
> 
> No, you would need a `javatoolkit-9999.ebuild` and point it to my fork or a local copy that contains my branch: https://github.com/the-horo/javatoolkit/tree/multi-release-jars. Here's the ebuild if it helps you:
> 
Seems to work, [my ebuild](https://github.com/gentoo/gentoo/compare/master...vaukai:javatoolkit) produces this output when emerging [xz-java](https://packages.gentoo.org/packages/dev-java/xz-java) (want to test with more ebuilds from [PR 40336](https://github.com/gentoo/gentoo/pull/40336) later today):
```
adding: org/tukaani/xz/simple/X86.class(in = 1473) (out= 1038)(deflated 29%)
adding: META-INF/versions/9/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/tukaani/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/tukaani/xz/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/tukaani/xz/check/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/tukaani/xz/check/CRC64.class(in = 1219) (out= 855)(deflated 29%)
adding: META-INF/versions/9/org/tukaani/xz/common/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/tukaani/xz/common/ByteArrayView.class(in = 2188) (out= 1081)(deflated 50%)
adding: META-INF/versions/9/org/tukaani/xz/lz/(in = 0) (out= 0)(stored 0%)
adding: META-INF/versions/9/org/tukaani/xz/lz/BasicMatchLengthFinder.class(in = 547) (out= 380)(deflated 30%)
adding: META-INF/versions/9/org/tukaani/xz/lz/MatchLength.class(in = 2236) (out= 1256)(deflated 43%)
adding: META-INF/versions/9/org/tukaani/xz/lz/MatchLengthFinder.class(in = 191) (out= 159)(deflated 16%)
adding: META-INF/versions/9/org/tukaani/xz/lz/UnalignedLongLEMatchLengthFinder.class(in = 725) (out= 499)(deflated 31%)
>>> Source compiled.
>>> Test phase [not enabled]: dev-java/xz-java-1.10

>>> Install dev-java/xz-java-1.10 into /var/tmp/portage/dev-java/xz-java-1.10/image
debug: java-pkg_dojar: entering function, parameters: xz-java.jar
debug: java-pkg_init_paths_: entering function, parameters: 
debug: JAVA_PKG_SHAREPATH: /usr/share/xz-java
debug: JAVA_PKG_ENV: /var/tmp/portage/dev-java/xz-java-1.10/image/usr/share/xz-java/package.env
debug: JAVA_PKG_JARDEST: /usr/share/xz-java/lib
debug: JAVA_PKG_LIBDEST: /usr/lib64/xz-java
debug: JAVA_PKG_WARDEST: /usr/share/xz-java/webapps
Good: 9 xz-java.jar META-INF/versions/9/module-info.class
Good: 1.8 xz-java.jar LZMADecDemo.class
Good: 1.8 xz-java.jar LZMAEncDemo.class
Good: 1.8 xz-java.jar TestAllocSpeed.class
Good: 1.8 xz-java.jar XZDecDemo.class
Good: 1.8 xz-java.jar XZEncDemo.class
Good: 1.8 xz-java.jar XZSeekDecDemo.class
Good: 1.8 xz-java.jar XZSeekEncDemo.class
Good: 1.8 xz-java.jar org/tukaani/xz/ARM64Options.class
Good: 1.8 xz-java.jar org/tukaani/xz/ARMOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/ARMThumbOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/ArrayCache.class
Good: 1.8 xz-java.jar org/tukaani/xz/BCJCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/BCJDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/BCJEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/BCJOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/BasicArrayCache$1.class
Good: 1.8 xz-java.jar org/tukaani/xz/BasicArrayCache$CacheMap.class
Good: 1.8 xz-java.jar org/tukaani/xz/BasicArrayCache$CyclicStack.class
Good: 1.8 xz-java.jar org/tukaani/xz/BasicArrayCache$LazyHolder.class
Good: 1.8 xz-java.jar org/tukaani/xz/BasicArrayCache.class
Good: 1.8 xz-java.jar org/tukaani/xz/BlockInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/BlockOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/CloseIgnoringInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/CorruptedInputException.class
Good: 1.8 xz-java.jar org/tukaani/xz/CountingInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/CountingOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/DeltaCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/DeltaDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/DeltaEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/DeltaInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/DeltaOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/DeltaOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/FilterCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/FilterDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/FilterEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/FilterOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/FinishableOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/FinishableWrapperOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/IA64Options.class
Good: 1.8 xz-java.jar org/tukaani/xz/IndexIndicatorException.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMA2Coder.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMA2Decoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMA2Encoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMA2InputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMA2Options.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMA2OutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMAInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/LZMAOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/MemoryLimitException.class
Good: 1.8 xz-java.jar org/tukaani/xz/PowerPCOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/RISCVOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/RawCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/ResettableArrayCache.class
Good: 1.8 xz-java.jar org/tukaani/xz/SPARCOptions.class
Good: 1.8 xz-java.jar org/tukaani/xz/SeekableFileInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/SeekableInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/SeekableXZInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/SimpleInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/SimpleOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/SingleXZInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/UncompressedLZMA2OutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/UnsupportedOptionsException.class
Good: 1.8 xz-java.jar org/tukaani/xz/X86Options.class
Good: 1.8 xz-java.jar org/tukaani/xz/XZ.class
Good: 1.8 xz-java.jar org/tukaani/xz/XZFormatException.class
Good: 1.8 xz-java.jar org/tukaani/xz/XZIOException.class
Good: 1.8 xz-java.jar org/tukaani/xz/XZInputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/XZOutputStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/check/CRC32.class
Good: 1.8 xz-java.jar org/tukaani/xz/check/CRC64.class
Good: 1.8 xz-java.jar org/tukaani/xz/check/Check.class
Good: 1.8 xz-java.jar org/tukaani/xz/check/None.class
Good: 1.8 xz-java.jar org/tukaani/xz/check/SHA256.class
Good: 1.8 xz-java.jar org/tukaani/xz/common/ByteArrayView.class
Good: 1.8 xz-java.jar org/tukaani/xz/common/DecoderUtil.class
Good: 1.8 xz-java.jar org/tukaani/xz/common/EncoderUtil.class
Good: 1.8 xz-java.jar org/tukaani/xz/common/StreamFlags.class
Good: 1.8 xz-java.jar org/tukaani/xz/common/Util.class
Good: 1.8 xz-java.jar org/tukaani/xz/delta/DeltaCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/delta/DeltaDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/delta/DeltaEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/index/BlockInfo.class
Good: 1.8 xz-java.jar org/tukaani/xz/index/IndexBase.class
Good: 1.8 xz-java.jar org/tukaani/xz/index/IndexDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/index/IndexEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/index/IndexHash.class
Good: 1.8 xz-java.jar org/tukaani/xz/index/IndexRecord.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/BT4.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/CRC32Hash.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/HC4.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/Hash234.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/LZDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/LZEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/MatchLength.class
Good: 1.8 xz-java.jar org/tukaani/xz/lz/Matches.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMACoder$LengthCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMACoder$LiteralCoder$LiteralSubcoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMACoder$LiteralCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMACoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMADecoder$1.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMADecoder$LengthDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMADecoder$LiteralDecoder$LiteralSubdecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMADecoder$LiteralDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMADecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoder$1.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoder$LengthEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoder$LiteralEncoder$LiteralSubencoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoder$LiteralEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoderFast.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/LZMAEncoderNormal.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/Optimum.class
Good: 1.8 xz-java.jar org/tukaani/xz/lzma/State.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeCoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeDecoderFromBuffer.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeDecoderFromStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeEncoderToBuffer.class
Good: 1.8 xz-java.jar org/tukaani/xz/rangecoder/RangeEncoderToStream.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/ARM.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/ARM64.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/ARMThumb.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/IA64.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/PowerPC.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/RISCVDecoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/RISCVEncoder.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/SPARC.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/SimpleFilter.class
Good: 1.8 xz-java.jar org/tukaani/xz/simple/X86.class
Good: 9 xz-java.jar META-INF/versions/9/org/tukaani/xz/check/CRC64.class
Good: 9 xz-java.jar META-INF/versions/9/org/tukaani/xz/common/ByteArrayView.class
Good: 9 xz-java.jar META-INF/versions/9/org/tukaani/xz/lz/BasicMatchLengthFinder.class
Good: 9 xz-java.jar META-INF/versions/9/org/tukaani/xz/lz/MatchLength.class
Good: 9 xz-java.jar META-INF/versions/9/org/tukaani/xz/lz/MatchLengthFinder.class
Good: 9 xz-java.jar META-INF/versions/9/org/tukaani/xz/lz/UnalignedLongLEMatchLengthFinder.class
CVV: 1.8
Checked: 137 Good: 137 Bad: 0 Skipped: 0
debug: java-pkg_append_: entering function, parameters: JAVA_PKG_CLASSPATH /usr/share/xz-java/lib/xz-java.jar
debug: installed xz-java.jar to /var/tmp/portage/dev-java/xz-java-1.10/image/usr/share/xz-java/lib
```

**Comment by vaukai** [comment]
Another thing is, in the whole source I cannot find any mention of `Multi-Release: true` which is crucial for `multi-release` as pointed out in the commit message of first commit in [this PR](https://github.com/gentoo/gentoo/pull/40336).

**Comment by the-horo** [comment]
Right, I'll add a check for it. If the line is not there should we skip the files under `META-INF/versions` or should we simply require that they are, at most, the version passed on the command line, like the regular `.class` files? I'm guessing the latter.

**Comment by vaukai** [comment]
> Right, I'll add a check for it. If the line is not there should we skip the files under `META-INF/versions` or should we simply require that they are, at most, the version passed on the command line, like the regular `.class` files? I'm guessing the latter.

For my understanding, it should error if `MANIFEST-MF` does not have `Multi-Release: true` while there are classes under `META-INF/versions`.
And also error if that line is there but no classes under `META-INF/versions`.
@fordfrog wdyt?

**Comment by the-horo** [comment]
A bunch of code changes, with tests and all but the behavior should stay the same.

> For my understanding, it should error if `MANIFEST-MF` does not have `Multi-Release: true` while there are classes under `META-INF/versions`. And also error if that line is there but no classes under `META-INF/versions`.

If there's no `Multi-Release: true` line in the manifest the code reports 1 error. If there is one but there are no `META-INF/versions` directories then it ignored. I could change it to your original suggestion if it's more desired.

Check `src/test/test_cvv.py` for details on how the code behaves in edge cases.

**Comment by fordfrog** [comment]
> If there's no `Multi-Release: true` line in the manifest the code reports 1 error. If there is one but there are no `META-INF/versions` directories then it ignored. I could change it to your original suggestion if it's more desired.

the multi release jar is still valid if `Multi-Release: true` is specified but no versioned classes are provided, so this way it should be ok.

**Comment by vaukai** [comment]
bug [900767](https://bugs.gentoo.org/900767) isn't yet tagged. 
Would  adding `[please reassign]` to the summary also helo for this repository?

**Comment by the-horo** [comment]
I'm guessing not as there wasn't any gentoo-bot message when I first made this PR and my original commit did contain the Closes tag. Though since you're the reporter you can add this PR to the See Also of the bug.

**Comment by vaukai** [comment]
Just updated [my ebuild ](https://github.com/gentoo/gentoo/compare/master...vaukai:javatoolkit) and tested again with several MR ebuilds, and I'd say: "It works."
The new check in fact reminded me of a missing `Multi-Release: true` which I finally [added](https://github.com/gentoo/gentoo/compare/fa275012f673a6bb1e06f4d504378ccded3a1fa7..25e8812b3474d557cf605c22b597e3e20fa7d51d) to the [log4j-api-2.24.3.ebuild](https://github.com/gentoo/gentoo/pull/40336).

> I'm guessing not as there wasn't any gentoo-bot message when I first made this PR and my original commit did contain the Closes tag. Though since you're the reporter you can add this PR to the See Also of the bug.

Ok. bug #40336 is for the ebuild and will be closed on removal of the old version.

@the-horo Thanks for doing this work for us.

File: src/javatoolkit/cvv.py (hunk 1 above)
Comment (vaukai) [reply src/javatoolkit/cvv.py]: > I'm not sure how this is relevant. If some version subdirectory in a JAR file doesn't contain a module-info the rest of the files would still be checked normally, i.e. that their bytecode encoded version is less or equal to the version of the subdirectory.
> 
> There is a bug however, the `file == "module-info.class"` checks is technically wrong as the filename in a multi-release jar would be something like `META-INF/versions/9/module-info.class` though practically the rest of the condition (`target_version < 9`) is false because `do_jar` skips versions less than 9.

We are building MR (multi-release) jar files only recently ([PR 40336](https://github.com/gentoo/gentoo/pull/40336) not yet merged), but `module-info.class` can be build in `::gentoo` only since commit [#4419bceb08](https://gitweb.gentoo.org/repo/gentoo.git/commit/eclass/java-pkg-simple.eclass?id=4419bceb0833e2b9ae045d9392f201ee359a84b7). 
Since that commit it was always placed directly in root of the jar file. Compiled with `target 9` but sitting in root even if other classes were built for `target 1.8`.
Binary blobs like `maven-bin` of course could have MR jar files before.

**Comment by the-horo** [comment]
@vaukai would you also be interested in changing the output format? In my *refactor* I stored a little bit more information but without printing it. Here's what i had in mind (above is the old format, below is the new one):
```
Good: 1.8 None A.class
Good: 1.8 mr.jar A.class
Good: 1.8 of.jar A.class
Good: 1.8 of.jar META-INF/versions/5/A.class
Good: 1.8 of.jar META-INF/versions/poof/A.class
Bad: 9 None B.class
Bad: 10 mr.jar META-INF/versions/10/A.class
Bad: 10 of.jar META-INF/versions/10/module-info.class
Bad: 10 of.jar META-INF/versions/10/A.class
Bad: 11 of.jar META-INF/versions/10/Module/A.class
Bad: 9 of.jar Module/B.class
Skipped: 9 of.jar module-info.class
CVV: 8
Checked: 12 Good: 5 Bad: 6 Skipped: 1
---------
Good: A.class version 1.8 (expected 1.8)
Good: mr.jar(A.class) version 1.8 (expected 1.8)
Good: of.jar(META-INF/versions/10/module-info.class) version 10 (expected 10)
Good: of.jar(A.class) version 1.8 (expected 1.8)
Good: of.jar(META-INF/versions/10/A.class) version 10 (expected 10)
Bad:  B.class version 9 (expected 1.8)
Bad:  mr.jar(META-INF/MANIFEST.MF) missing "Multi-Release: true" implied by ['META-INF/versions/10']
Bad:  of.jar(META-INF/versions/10/Module/A.class) version 11 (expected 10)
Bad:  of.jar(Module/B.class) version 9 (expected 1.8)
Skip: of.jar(module-info.class) because: A module-info requires java release >= 9
Skip: of.jar(META-INF/versions/5) because: The version directory "5" is less than 9
Skip: of.jar(META-INF/versions/poof) because: The version directory "poof" is not a number
Checked: 12 Good: 5 Bad: 4 Skipped: 3
```

I can change it to something else if you have suggestions

**Comment by vaukai** [comment]
> @vaukai would you also be interested in changing the output format? In my _refactor_ I stored a little bit more information but without printing it. Here's what i had in mind (above is the old format, below is the new one):
> 
> ```
> Good: 1.8 None A.class
> Good: 1.8 mr.jar A.class
> Good: 1.8 of.jar A.class
> Good: 1.8 of.jar META-INF/versions/5/A.class
> Good: 1.8 of.jar META-INF/versions/poof/A.class
> Bad: 9 None B.class
> Bad: 10 mr.jar META-INF/versions/10/A.class
> Bad: 10 of.jar META-INF/versions/10/module-info.class
> Bad: 10 of.jar META-INF/versions/10/A.class
> Bad: 11 of.jar META-INF/versions/10/Module/A.class
> Bad: 9 of.jar Module/B.class
> Skipped: 9 of.jar module-info.class
> CVV: 8
> Checked: 12 Good: 5 Bad: 6 Skipped: 1
> ---------
> Good: A.class version 1.8 (expected 1.8)
> Good: mr.jar(A.class) version 1.8 (expected 1.8)
> Good: of.jar(META-INF/versions/10/module-info.class) version 10 (expected 10)
> Good: of.jar(A.class) version 1.8 (expected 1.8)
> Good: of.jar(META-INF/versions/10/A.class) version 10 (expected 10)
> Bad:  B.class version 9 (expected 1.8)
> Bad:  mr.jar(META-INF/MANIFEST.MF) missing "Multi-Release: true" implied by ['META-INF/versions/10']
> Bad:  of.jar(META-INF/versions/10/Module/A.class) version 11 (expected 10)
> Bad:  of.jar(Module/B.class) version 9 (expected 1.8)
> Skip: of.jar(module-info.class) because: A module-info requires java release >= 9
> Skip: of.jar(META-INF/versions/5) because: The version directory "5" is less than 9
> Skip: of.jar(META-INF/versions/poof) because: The version directory "poof" is not a number
> Checked: 12 Good: 5 Bad: 4 Skipped: 3
> ```
> 
> I can change it to something else if you have suggestions

I'm fully satisfied with the old format. But let's hear how @fordfrog likes your proposal.

**Comment by fordfrog** [comment]
i'd say the second format contains some additional info which might be useful to understand and resolve encountered issues so i would prefer the enhanced output.

**Comment by the-horo** [comment]
If we can't decide on an output format let's keep both and avoid the otherwise useful changes getting stalled.

I've kept the default format the one from the current version, the new format can be enabled by passing `--detailed-output`

**Comment by vaukai** [comment]
> If we can't decide on an output format let's keep both and avoid the otherwise useful changes getting stalled.
> 
> I've kept the default format the one from the current version, the new format can be enabled by passing `--detailed-output`

I am sure I agreed to fordfrog's [comment](https://github.com/gentoo/javatoolkit/pull/3#issuecomment-2675297121) but obviously my response got lost.

**Comment by the-horo** [comment]
> I am sure I agreed to fordfrog's [comment](https://github.com/gentoo/javatoolkit/pull/3#issuecomment-2675297121) but obviously my response got lost.

It happens sometimes.

I've unconditionally changed the format to what I proposed and I've also listed the relevant changes in the `NEWS` file.

**Review by arthurzam — APPROVED** [review]

**Review by vaukai — APPROVED** [review]
Approved
